### PR TITLE
Cosmetic suggestions

### PR DESCRIPTION
**Pull request by damithc**
Did some tweaks to the code. I'll add comments to explain the reasons.

### CODE REVIEW

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 1]
```diff
@@ -1,26 +1,35 @@
...
- * in a text file. Each person has a name, a phone number (p/) and an email (e/).
- *
- * Furthermore, you can search for persons by name using keywords.
- * When multiple items are displayed, they are ordered by insertion time.
- *
- * Further command information can be found in the README or by entering 'help' as input to the program.
- * The program takes 1 argument at the start: the path of the storage file.
- * If no arguments are found, it uses a default storage file path.
+ * This class is used to maintain a list of person data which are saved
+ * in a text file.
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Let's put all usage info in README instead of here.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 2]
```diff
@@ -1,26 +1,35 @@
...
 
+    /**
+     * Default file path used if the user doesn't provide the file name.
+     */
+    private static final String DEFAULT_STORAGE_FILEPATH = "storage.txt";
+
+    /**
+     * A platform independent line separator.
+     */
+    private static final String LS = System.lineSeparator();
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: It's a good practice not to use `\n` as it doesn't translate equally to Unix

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 3]
```diff
@@ -30,188 +39,258 @@
...
+
     /**
-     * Stores the last viewed person listing (from commands like find and list)
-     * Commands using an index to target a person will refer to the indices in this list.
-     *
-     * Persons in this list are the same objects as the corresponding elements in the backing model.
-     * Deleting persons in the backing model may not necessarily update this latest view snapshot.
+     * Stores the most recent list of persons shown to the user as a result of a user command.
+     * This is a subset of the full list. Deleting persons in the pull list does not delete
+     * those persons from this list.
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Tried to avoid terms such as model, backing model etc.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 4]
```diff
@@ -30,188 +39,258 @@
...
-        while (!isExitRequested) {
-            latestUserInput = getUserInput();
-            String feedback = parseAndExecuteCommand(latestUserInput);
+        loadDataFromStorage();
+        while (true) {
+            userCommand = getUserInput();
+            echoUserCommand(userCommand);
+            String feedback = executeCommand(userCommand);
             showToUser(feedback);
         }
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Went back to `while(true)` loop because it was not clear from this code where the flag is being set.  This method should be very high level and self-explanatory.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 5]
```diff
@@ -30,188 +39,258 @@
...
+
+        if (args.length == 1) {
+            setupGivenFileForStorage(args[0]);
+        }
+
+        if(args.length == 0) {
+            setupDefaultFileForStorage();
+        }
+
     }
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Rejigged this method to make it more cohesive and SLAP. Found a but in the process.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 6]
```diff
@@ -30,188 +39,258 @@
...
     /**
-     * Informs user if program args are invalid.
-     *
-     * @param args full program run arguments passed to application main method
+     * Sets up the storage file based on the supplied file path.
+     * Creates the file if it is missing.
+     * Exits if the file name is not acceptable.
      */
-    private static void informUserIfProgramArgsAreInvalid(String[] args) {
-        if (args.length > 2) {
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: missing `=` here

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 7]
```diff
@@ -30,188 +39,258 @@
...
+        showToUser(MESSAGE_USING_DEFAULT_FILE);
+        storageFilePath = DEFAULT_STORAGE_FILEPATH;
+        createFileIfMissing(storageFilePath);
     }
 
     /**
-     * Cleans up all loose ends, and prepares program for termination.
+     * Returns true if the given file is acceptable.
+     * The file path is acceptable if it ends in '.txt'
+     * TODO: Implement a more rigorous validity checking.
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: No need to do this now, but we should document the behavior in the usage info

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 8]
```diff
@@ -475,32 +552,38 @@ private static String getUserInput() {
...
+     * Shows the list of persons to the user.
+     * The list will be indexed, starting from 1.
      *
-     * @param persons to show
      */
     private static void showToUser(List<String[]> persons) {
+        String listAsString = getDisplayString(persons);
+        showToUser(listAsString);
+        updateLatestViewedPersonListing(persons);
+    }
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Abstracted a code chunk to make this method SLAP

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 9]
```diff
@@ -557,53 +641,60 @@ private static void updateLatestViewedPersonListing(List<String[]> newListing) {
...
-     *         problems decoding: empty optional
+     * @return the list of decoded persons
      */
-    private static Optional<List<String[]>> loadPersonsFromFile(String filePath) {
-        try (final BufferedReader storageReader =
-                     new BufferedReader(new FileReader(filePath))) {
-            final List<String[]> decodedPersons = new ArrayList<>();
+    private static List<String[]> loadPersonsFromFile(String filePath) {
+        final List<String[]> decodedPersons = new ArrayList<>();
+        //TODO: separate file reading from decoding
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Can be two methods.
1. Read a file into a List<String>
2. Convert List<String> to List<String[ ]>

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 10]
```diff
@@ -557,53 +641,60 @@ private static void updateLatestViewedPersonListing(List<String[]> newListing) {
...
-     * Converts contents of a file into a list of persons. Shows feedback to user.
+     * Converts contents of a file into a list of persons.
+     * Exits program if any errors in reading or decoding was encountered.
      *
      * @param filePath file to load from
-     * @return able to decode: optional containing list of persons in file
-     *         problems decoding: empty optional
+     * @return the list of decoded persons
      */
-    private static Optional<List<String[]>> loadPersonsFromFile(String filePath) {
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: As we exist upon error, no need to use Optional here.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 11]
```diff
@@ -699,16 +791,6 @@ private static void clearAddressBook() {
...
-    /**
-     * Resets the internal model with the given data. Does not save to file.
-     *
-     * @param persons list of persons to initialise the model with
-     */
-    private static void initialiseAddressBookModel(List<String[]> persons) {
-        ALL_PERSONS.clear();
-        ALL_PERSONS.addAll(persons);
-    }
-
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: This could be inlined without losing SLAP.

File: test/input.txt [hunk 12]
```diff
@@ -0,0 +1,31 @@
+#invalid command
+sfdfd
+add Adam Brown p/111111 e/adam@gmail.com
```
Comment (damithc) [inline test/input.txt]: Added some prelim test cases. need to be expanded.

File: test/input.txt [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+list
+add Betsy Choo p/222222 e/benchoo@nus.edu.sg
+list
+add Charlie Dickson p/33333333 e/charlie.d@nus.edu.sg
+list
+add Dickson Ee p/444444 e/dickson@nus.edu.sg
+list
+find
+find bet
+find betsy
```
Comment (damithc) [inline test/input.txt]: Our search is case sensitive. Should be documented in usage info.

**Comment by damithc** [comment]
@YijinL for your review and merge. Add the missing stuff. We should move to level 2 on Monday.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 14]
```diff
@@ -887,6 +974,7 @@ private static boolean isPersonDataValid(String name, String phone, String email
      */
     private static boolean isPersonNameValid(String name) {
         return name.matches("(\\w|\\s)+");  // name is nonempty mixture of alphabets and whitespace
+        //TODO: implement a more permissive validation
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: No need to do now

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 11 above)
Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: This means other places can call `ALL_PERSONS` `List` methods? I worry that students will see it and add a new CRUD feature which changes the master list directly, and forget to trigger a save to file.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 15]
```diff
@@ -30,188 +39,258 @@
...
+    private static final char INPUT_COMMENT_MARKER = '#';
 
     /*
      * This variable is declared for the whole class (instead of declaring it
      * inside the readUserCommand() method to facilitate automated testing using
      * the I/O redirection technique. If not, only the first line of the input
-     * text file will be processed. TODO FIND OUT WHY
+     * text file will be processed.
      */
     private static final Scanner SCANNER = new Scanner(System.in);
```
Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: I am thinking of shifting the scanner into the `main` method, and passing it as an argument to `readUserInput`. This way other methods cannot mess with the scanner.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 16]
```diff
@@ -30,188 +39,258 @@
...
+        showToUser(MESSAGE_WELCOME);
+    }
+
+    /*
+     * ==============NOTE TO STUDENTS======================================
+     * Parameter description can be omitted from the method header comment
+     * if the parameter name is self-explanatory.
+     * In the method below, '@param userInput' comment has been omitted.
+     * ====================================================================
+     */
```
Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: How about simple pure functions (no side effects)? Like getters `getX()` and predicates `isX(testee)`. The header comment can be merged into the `@return` tag body.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 17]
```diff
@@ -30,188 +39,258 @@
...
+    /**
+     * Echoes the user input back to the user.
+     */
+    private static void echoUserCommand(String userCommand) {
+        showToUser(System.lineSeparator() + "[Command entered:" + userCommand + "]");
```
Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: Can use `LS` instead.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 18]
```diff
@@ -475,32 +552,38 @@ private static String getUserInput() {
...
     private static void showToUser(List<String[]> persons) {
+        String listAsString = getDisplayString(persons);
+        showToUser(listAsString);
+        updateLatestViewedPersonListing(persons);
+    }
+
+    /**
+     * Returns the display string representation of the list of persons.
+     */
+    private static String getDisplayString(List<String[]> persons) {
```
Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: Maybe name should indicate that the result person strings are 1-indexed

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 11 above)
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: I see. Can revert back to original.

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 15 above)
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: There are lot of other data that are declared at class level so trying to protect scanner specifically doesn't make much sense. Putting it in the main method hurts the abstraction level of the main method (as scanner is a low level detail). Ideally, it should hidden inside the readUserInput method but we can't do that because it will not work for our test technique.

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 16 above)
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Based on Google style guide, it's preferred to have the overview sentence and omit other stuff than the other way around. The reason given is, overview sentence is the only thing that is shown in some contexts.  So I'd rather have this
```
/** Returns ...
```
than this
```
/** @return ...
```

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 17 above)
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: yes, I missed that.

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 18 above)
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Sure. The word 'display' implies it is 'one-indexed' (because it is known that display is one-indexed)?

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 9 above)
Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: For 2 : `List<String> -> Optional<List<String[]>>`,
Separating into it's own side-effect-less `decodePersonsFromStrings` method means that the error message with the corrupted line number and contents cannot be constructed (no access to those internal fields). 

Maybe we should keep the decoding logic inside this method for now, and separate it for the OO version, using exceptions to make the detailed error message?

Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: If step 1 converts the file in to a List<String>, you should still be able to know line numbers, right? ArrayList preserves the order. So get(0) will give you the first line in the file, and so on.

Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: But the decoding method returns an empty optional if any of the given strings cannot be decoded, and the empty optional does not carry information of the offending line number or contents.

Even if we change the decoding method to return `List<Optional<String[]>>`, it also does not contain information of the offending line contents.

File: README.md [hunk 19]
```diff
@@ -6,8 +6,8 @@
```
Comment (YijinL) [inline README.md]: @damithc learning objectives

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 9 above)
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Why doesn't this work?
```java
private static List<String[]> loadPersonsFromFile(String filePath) {
        final List<String[]> decodedPersons = new ArrayList<>();
        List<String> lines = getLinesFromFile(filePath);
        for (int i = 0; i < lines.size(); i++) {
            String line = lines.get(i);
            final Optional<String[]> successfullyDecodedPerson = decodePersonFromStringRepresentation(line);

            // unable to decode person means file content format invalid; stop program
            if (!successfullyDecodedPerson.isPresent()) {
                showToUser(getMessageForInvalidPersonLineInFile(filePath, i, line));
                exitProgram();
            }

            decodedPersons.add(successfullyDecodedPerson.get());
        }
        return decodedPersons;
    }
```

File: README.md (hunk 19 above)
Comment (damithc) [inline README.md]: Sorry, didn't get what you mean.

**Comment by YijinL** [comment]
Finished as far as I can tell. @damithc review
- `README.md` left with learning objectives
- More exhaustive testing
- Tests now properly grouped, commented

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 9 above)
Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: It works, but I cannot extract the {decode list of strings into list of persons} as a separate method because there is no nice way to pass exception information back for user feedback.

File: README.md (hunk 19 above)
Comment (YijinL) [inline README.md]: Module coordinator(s) should fill this list :P

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 9 above)
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Understood. Do what you can. The current one inter-twine two things. I still think you can write something like this.
```java
private static List<String[]> loadPersonsFromFile(String filePath) {
        List<String> lines = getLinesFromFile(filePath);
        return convertToPersonData(lines);
}
```
The only piece of information `convertToPersonData` is missing is the file path, which can be passed as a parameter or accessed from the static variable.

Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: In the latest commit, I added these twin methods to the `PERSON METHODS` section:
```java
   /**
     * Encodes list of persons into list of decodable and readable string representations.
     *
     * @param persons to be encoded
     * @return encoded strings
     */
    private static List<String> encodePersonsToStrings(List<String[]> persons) {
        final List<String> encoded = new ArrayList<>();
        for (String[] person : persons) {
            encoded.add(encodePersonToString(person));
        }
        return encoded;
    }

    /**
     * Decode persons from a list of string representations.
     *
     * @param encodedPersons strings to be decoded
     * @return if cannot decode any: empty Optional
     *         else: Optional containing decoded persons
     */
    private static Optional<List<String[]>> decodePersonsFromStrings(List<String> encodedPersons) {
        final List<String[]> decodedPersons = new ArrayList<>();
        for (String encodedPerson : encodedPersons) {
            final Optional<String[]> decodedPerson = decodePersonFromString(encodedPerson);
            if (!decodedPerson.isPresent()) {
                return Optional.empty();
            }
            decodedPersons.add(decodedPerson.get());
        }
        return Optional.of(decodedPersons);
    }
```
Here the decode method does a pure conversion with no side effects; it might be weird for a `convertPersonToLines` method to exit the program if a corrupted line is encountered.

File: README.md (hunk 19 above)
Comment (damithc) [inline README.md]: Yes, will do :-)

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 9 above)
Comment (YijinL) [inline src/nus/cs2103/addressbook/AddressBook.java]: Maybe we can remove the super detailed error message
```java
private static final String MESSAGE_INVALID_PERSON_LINE_IN_FILE = "Storage file contents invalid:"
                                                               + LS + "\tFile: %1$s\tLine %2$d:"
                                                               + LS + "\tNot a valid person:\t%3$s";
```
And just show a simpler message. If we do this the whole problem goes away, and we can leave the detailed error feedback as a student exercise?

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 20]
```diff
@@ -557,53 +644,58 @@ private static void updateLatestViewedPersonListing(List<String[]> newListing) {
...
-            int lineNumber = 1;
-
-            // do for each line in the file, also tracks line number
-            for (String line = storageReader.readLine(); line != null; line = storageReader.readLine(), lineNumber++) {
-                final Optional<String[]> successfullyDecodedPerson = decodePersonFromStringRepresentation(line);
+     * @return the list of decoded persons
+     */
+    private static List<String[]> loadPersonsFromFile(String filePath) {
+        final List<String[]> decodedPersons = new ArrayList<>();
+        try {
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: This try/catch block should be in the method that reads the file, not in this method.

File: src/nus/cs2103/addressbook/AddressBook.java [hunk 21]
```diff
@@ -557,53 +644,58 @@ private static void updateLatestViewedPersonListing(List<String[]> newListing) {
...
+            return;
+        }
+
+        showToUser(String.format(MESSAGE_ERROR_MISSING_STORAGE_FILE, filePath));
+
         try {
-            if (storageFile.createNewFile()) {
-                showToUser(String.format(MESSAGE_ERROR_MISSING_STORAGE_FILE, filePath));
-                showToUser(String.format(MESSAGE_STORAGE_FILE_CREATED, filePath));
-            }
```
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Forgot to comment why I changed this. I try to follow the command-query separation principle when I can. When a command operation (i.e. not a query operation) returns the result as the return value, using that value in the code makes the call site code hard to follow. e.g. `if (storageFile.createNewFile())` is not very intuitive. Even Java doesn't follow that principle, so it can't be all that bad. But whenever we can, we should try to follow it.

**Comment by damithc** [comment]
Looks good. A few comments added. May be the header comment of the find method should document its exact behavior. i.e. OR search, case sensitive

**Comment by damithc** [comment]
Can merge after fixing the pending issues. After that I'll add learning outcomes and the exercises.

File: src/nus/cs2103/addressbook/AddressBook.java (hunk 9 above)
Comment (damithc) [inline src/nus/cs2103/addressbook/AddressBook.java]: Fair point about low level methods doing exits. However, we do exits all over the place anyway. And that can be used as the motivation for introducing exception handling at a later stage. 
I don't mind a less detailed error message, for simplicity.
